File: src/main/scala/io/vertx/ext/asyncsql/impl/AsyncSQLClientImpl.scala
package io.vertx.ext.asyncsql.impl

import io.vertx.core.{AsyncResult, Handler, Vertx}
import io.vertx.core.json.JsonObject
import io.vertx.ext.asyncsql.AsyncSQLClient
import io.vertx.ext.sql.SQLConnection

/**
 * @author <a href="http://www.campudus.com">Joern Bernhardt</a>.
 */
class AsyncSQLClientImpl(vertx: Vertx, config: JsonObject, mysql: Boolean) extends AsyncSQLClient {

  val baseService: BaseSQLClient = {
    if (mysql) {
      new MySQLClient(vertx, config)
    } else {
      new PostgreSQLClient(vertx, config)
    }
  }

  override def close(whenDone: Handler[AsyncResult[Void]]): Unit = baseService.close(whenDone)

  override def close(): Unit = baseService.close(null)

  override def getConnection(handler: Handler[AsyncResult[SQLConnection]]): Unit = {
    baseService.getConnection(handler)
  }
}

File: src/main/java/io/vertx/ext/asyncsql/impl/AsyncSQLClientImpl.java
package io.vertx.ext.asyncsql.impl;

import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.asyncsql.AsyncSQLClient;
import io.vertx.ext.sql.SQLConnection;

/**
 * @author <a href="http://www.campudus.com">Joern Bernhardt</a>.
 */
public class AsyncSQLClientImpl implements AsyncSQLClient {

  private final BaseSQLClient baseClient;

  public AsyncSQLClientImpl(Vertx vertx, JsonObject config, boolean mysql) {
    if (mysql) {
      baseClient = new MYSQLClientImpl(vertx, config);
    } else {
      baseClient = new PostgreSQLClientImpl(vertx, config);
    }
  }

  @Override
  public void close() {
    baseClient.close(null);
  }

  @Override
  public void close(Handler<AsyncResult<Void>> whenDone) {
    baseClient.close(whenDone);
  }

  @Override
  public void getConnection(Handler<AsyncResult<SQLConnection>> handler) {
    baseClient.getConnection(handler);
  }

}
